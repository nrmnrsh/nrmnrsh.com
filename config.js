System.config({
  baseURL: "/",
  defaultJSExtensions: true,
  transpiler: "babel",
  babelOptions: {
    "optional": [
      "runtime",
      "optimisation.modules.system"
    ]
  },
  paths: {
    "app/*": "sources/js/src/*.js",
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "babel": "npm:babel-core@5.8.38",
    "babel-runtime": "npm:babel-runtime@5.8.38",
    "backbone": "npm:backbone@1.3.3",
    "backbone.geppetto": "npm:backbone.geppetto@0.7.1",
    "core-js": "npm:core-js@1.2.7",
    "es5shim": "github:es-shims/es5-shim@4.5.2",
    "i18next": "github:i18next/i18next@5.0.0",
    "jquery": "github:components/jquery@2.2.4",
    "jquery-qunit": "github:qunitjs/qunit@1.20.0",
    "lolex": "github:sinonjs/lolex@1.4.0",
    "picnic": "github:moccu/picnic@0.7.2",
    "sinon": "github:sinonjs/sinon@1.17.3",
    "sinon-qunit": "github:cjohansen/sinon-qunit@2.0.0",
    "text": "github:systemjs/plugin-text@0.0.2",
    "underscore": "npm:underscore@1.8.3",
    "github:jspm/nodelibs-assert@0.1.0": {
      "assert": "npm:assert@1.4.1"
    },
    "github:jspm/nodelibs-buffer@0.1.1": {
      "buffer": "npm:buffer@5.0.6"
    },
    "github:jspm/nodelibs-path@0.1.0": {
      "path-browserify": "npm:path-browserify@0.0.0"
    },
    "github:jspm/nodelibs-process@0.1.2": {
      "process": "npm:process@0.11.10"
    },
    "github:jspm/nodelibs-util@0.1.0": {
      "util": "npm:util@0.10.3"
    },
    "github:jspm/nodelibs-vm@0.1.0": {
      "vm-browserify": "npm:vm-browserify@0.0.4"
    },
    "github:moccu/picnic@0.7.2": {
      "backbone": "npm:backbone@1.3.3",
      "backbone.geppetto": "npm:backbone.geppetto@0.7.1",
      "jquery": "github:components/jquery@2.2.4",
      "kenwheeler/slick": "github:kenwheeler/slick@1.6.0",
      "text": "github:systemjs/plugin-text@0.0.2",
      "underscore": "npm:underscore@1.8.3"
    },
    "npm:assert@1.4.1": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "buffer": "github:jspm/nodelibs-buffer@0.1.1",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "util": "npm:util@0.10.3"
    },
    "npm:babel-runtime@5.8.38": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:backbone.geppetto@0.7.1": {
      "backbone": "npm:backbone@1.3.3",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "underscore": "npm:underscore@1.8.3"
    },
    "npm:backbone@1.3.3": {
      "process": "github:jspm/nodelibs-process@0.1.2",
      "underscore": "npm:underscore@1.8.3"
    },
    "npm:buffer@5.0.6": {
      "base64-js": "npm:base64-js@1.2.1",
      "ieee754": "npm:ieee754@1.1.8"
    },
    "npm:core-js@1.2.7": {
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "path": "github:jspm/nodelibs-path@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "systemjs-json": "github:systemjs/plugin-json@0.1.2"
    },
    "npm:inherits@2.0.1": {
      "util": "github:jspm/nodelibs-util@0.1.0"
    },
    "npm:path-browserify@0.0.0": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:process@0.11.10": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "vm": "github:jspm/nodelibs-vm@0.1.0"
    },
    "npm:util@0.10.3": {
      "inherits": "npm:inherits@2.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:vm-browserify@0.0.4": {
      "indexof": "npm:indexof@0.0.1"
    }
  }
});
