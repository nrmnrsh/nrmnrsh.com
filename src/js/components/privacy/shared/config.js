export const EVENT_CHANGE = 'privacy:change';
export const EVENT_VISIBILITY = 'privacy:visibility';

export const NAMESPACE_MODEL = 'privacy:consents';
export const NAMESPACE_VIEW = 'privacy:views';

export const CONSENT_ANALYTICS = 'analytics';
export const CONSENT_LOGROCKET = 'logrocket';
