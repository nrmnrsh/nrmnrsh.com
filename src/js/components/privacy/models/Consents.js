import { Model } from 'pacto.model';

import {CONSENT_ANALYTICS, CONSENT_LOGROCKET} from 'components/privacy/shared/config';


const STORAGE_KEY = 'nrmnrsh:privacy:consents:v1';

export class Consents extends Model {

	get defaults() {
		return {
			[CONSENT_ANALYTICS]: undefined,
			[CONSENT_LOGROCKET]: undefined,
		};
	}

	get hasMissing() {
		return Object.values(this.props)
			.filter((consent) => consent === undefined)
			.length > 0;
	}

	load() {
		try {
			const data = window.localStorage.getItem(STORAGE_KEY);
			const parsed = JSON.parse(data) || {};
			Object.entries(parsed)
				.forEach(([key, value]) => {
					this.props[key] = value;
				});
		} catch(error) {
			window.console.error('Consent load error', error);
		}
	}

	save() {
		try {
			const data = JSON.stringify(this.props);
			window.localStorage.setItem(STORAGE_KEY, data);
		} catch(error) {
			window.console.error('Consent save error', error);
		}
	}

}
