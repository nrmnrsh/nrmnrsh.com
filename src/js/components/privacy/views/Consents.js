import {View} from 'pacto';

import {CONSENT_ANALYTICS, CONSENT_LOGROCKET, EVENT_CHANGE} from 'components/privacy/shared/config';


const SELECTOR_DIALOG ='[role="dialog"]';
const SELECTOR_CHECKBOXES = '[type="checkbox"][name="privacy"]';
const SELECTOR_BUTTON_ALLOW_ALL = '.allow-all';
const SELECTOR_BUTTON_ALLOW_SELECTION = '.allow-selection';

const TEMPLATE = `
	<div
		role="dialog"
		aria-modal="true"
		aria-hidden="true"
		aria-labelledby="privacy-title"
		aria-describedby="privacy-description"
	>
		<h2 id="privacy-title">Enable services</h2>
		<div id="privacy-description">
			<div class="description">
				<p>
					This site is using different services that allow me to enhance and
					analyze the usage of this page. They might uses cookies to work properly...
				</p>
				<button role="button" class="allow-all">Enable all</button>
			</div>

			<details>
				<summary>Customize services</summary>

				<p>
					I believe preventing cookie usage is not the only way to protect
					your privacy. Instead, you will have full control over all the services
					that are loaded to control any data exchange to foreign services.
				</p>

				<form class="details">
					<dl>
						<dt>
							<label>
								<input type="checkbox" name="privacy" value="${CONSENT_ANALYTICS}" />
								Google Analytics
							</label>
						</dt>
						<dd>
							Google Analytics is a web analytics service offered by Google that
							tracks and reports website traffic. I track page views and clicks
							on links. The service is running using the "anonymizeIp" setting.
							Read more about <a
								href="https://policies.google.com/privacy?hl=en#footnote-link-info"
								rel="noopener noreferrer"
								target="_blank"
							>data privacy</a>
							from Google Analytics.
						</dd>

						<dt>
							<label>
								<input type="checkbox" name="privacy" value="${CONSENT_LOGROCKET}" />
								Logrocket
							</label>
						</dt>
						<dd>
							LogRocket is a cloud-based and on-premise logging and session replay
							platform for JavaScript applications, which helps me track UX problems
							and analyze as well as remediate the root causes of bugs.
							Read more about <a
								href="https://logrocket.com/privacy/"
								rel="noopener noreferrer"
								target="_blank"
							>privacy policy</a>
							from LogRocket.
						</dd>
					</dl>

					<button role="button" class="allow-selection">Use selection</button>
				</form>
			</details>
		</div>
	</div>
`;


export class Consents extends View {

	constructor(options) {
		super(options);
		this._model = options.model;

		this._onAllowAll = this._onAllowAll.bind(this);
		this._onAllowSelection = this._onAllowSelection.bind(this);
		this._onConsentChange = this._onConsentChange.bind(this);
	}

	get visible() {
		return this._dialog.getAttribute('aria-hidden') !== 'true';
	}

	set visible(value) {
		this._dialog.setAttribute('aria-hidden', !value);
	}

	render() {
		const {el} = this;

		el.innerHTML = TEMPLATE;

		this._model.on('change', this._onConsentChange);
		this._dialog = el.querySelector(SELECTOR_DIALOG);
		this._checkboxes = [...el.querySelectorAll(SELECTOR_CHECKBOXES)];

		this._buttonAll = el.querySelector(SELECTOR_BUTTON_ALLOW_ALL);
		this._buttonAll.addEventListener('click', this._onAllowAll);
		this._buttonSelection = el.querySelector(SELECTOR_BUTTON_ALLOW_SELECTION);
		this._buttonSelection.addEventListener('click', this._onAllowSelection);

		this.update();
		return this;
	}

	update() {
		this._checkboxes.forEach((checkbox) => {
			const name = checkbox.value;
			checkbox.checked = !!this._model.props[name];
		});
	}

	_onAllowAll(event) {
		event.preventDefault();
		const props = {};

		this._checkboxes.forEach((checkbox) => {
			const name = checkbox.value;
			props[name] = true;
		});

		this.context.trigger(EVENT_CHANGE, {props});
	}

	_onAllowSelection(event) {
		event.preventDefault();
		const props = {};

		this._checkboxes.forEach((checkbox) => {
			const name = checkbox.value;
			props[name] = checkbox.checked;
		});

		this.context.trigger(EVENT_CHANGE, {props});
	}

	_onConsentChange() {
		this.update();
	}

}
