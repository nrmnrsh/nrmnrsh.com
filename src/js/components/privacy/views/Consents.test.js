import {Context} from 'pacto';

import {Consents as Model} from 'components/privacy/models/Consents';
import {CONSENT_ANALYTICS, CONSENT_LOGROCKET, EVENT_CHANGE} from 'components/privacy/shared/config';

import {Consents as View} from './Consents';


describe('The privacy consents view', () => {

	let el;
	let model
	let context;

	beforeEach(() => {
		el = document.createElement('div');
		model = new Model();
		context = new Context();
	});

	afterEach(() => {
		el = null;
		context = null;
	});

	it('should render', () => {
		const view = new View({ el, context, model });
		view.render();

		expect(view.el)
			.toMatchSnapshot();
	});

	it('should initially be hidden', () => {
		const view = new View({ el, context, model });
		view.render();

		expect(view.visible)
			.toBeFalsy();
		expect(view.el.firstElementChild.getAttribute('aria-hidden'))
			.toBe('true');
	});

	it('should change all props as true when click on "allow all" button', () => {
		const callback = jest.fn();
		context.on(EVENT_CHANGE, callback);

		const view = new View({ el, context, model });
		view.render();

		const button = view.el.querySelector('.allow-all');
		const event = document.createEvent('MouseEvent');
		event.initMouseEvent('click');

		button.dispatchEvent(event);
		expect(callback)
			.toHaveBeenCalledTimes(1);
		expect(callback)
			.toHaveBeenCalledWith(
				expect.objectContaining({
					sender: context,
					data: {
						props: {
							[CONSENT_ANALYTICS]: true,
							[CONSENT_LOGROCKET]: true,
						},
					},
				}),
			);
	});

	it('should change all props when click on "allow selection" button', () => {
		const callback = jest.fn();
		context.on(EVENT_CHANGE, callback);

		const view = new View({ el, context, model });
		view.render();

		const analytics = view.el.querySelector(`input[value="${CONSENT_ANALYTICS}"]`);
		analytics.checked = true;

		const logrocket = view.el.querySelector(`input[value="${CONSENT_LOGROCKET}"]`);
		logrocket.checked = false;

		const button = view.el.querySelector('.allow-selection');
		const event = document.createEvent('MouseEvent');
		event.initMouseEvent('click');

		button.dispatchEvent(event);
		expect(callback)
			.toHaveBeenCalledTimes(1);
		expect(callback)
			.toHaveBeenCalledWith(
				expect.objectContaining({
					sender: context,
					data: {
						props: {
							[CONSENT_ANALYTICS]: true,
							[CONSENT_LOGROCKET]: false,
						},
					},
				}),
			);
	});

	it('should update checkboxes when model props changes', () => {
		const view = new View({ el, context, model });
		view.render();

		const analytics = view.el.querySelector(`input[value="${CONSENT_ANALYTICS}"]`);
		const logrocket = view.el.querySelector(`input[value="${CONSENT_LOGROCKET}"]`);

		model.props[CONSENT_ANALYTICS] = true;
		expect(analytics.checked).toBeTruthy();
		expect(logrocket.checked).toBeFalsy();

		model.props[CONSENT_ANALYTICS] = false;
		expect(analytics.checked).toBeFalsy();
		expect(logrocket.checked).toBeFalsy();

		model.props[CONSENT_LOGROCKET] = false;
		expect(analytics.checked).toBeFalsy();
		expect(logrocket.checked).toBeFalsy();

		model.props[CONSENT_LOGROCKET] = true;
		expect(analytics.checked).toBeFalsy();
		expect(logrocket.checked).toBeTruthy();
	});

});