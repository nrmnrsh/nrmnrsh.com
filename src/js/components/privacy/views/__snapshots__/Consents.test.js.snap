// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The privacy consents view should render 1`] = `
<div>
  
	
  <div
    aria-describedby="privacy-description"
    aria-hidden="true"
    aria-labelledby="privacy-title"
    aria-modal="true"
    role="dialog"
  >
    
		
    <h2
      id="privacy-title"
    >
      Enable services
    </h2>
    
		
    <div
      id="privacy-description"
    >
      
			
      <div
        class="description"
      >
        
				
        <p>
          
					This site is using different services that allow me to enhance and
					analyze the usage of this page. They might uses cookies to work properly...
				
        </p>
        
				
        <button
          class="allow-all"
          role="button"
        >
          Enable all
        </button>
        
			
      </div>
      

			
      <details>
        
				
        <summary>
          Customize services
        </summary>
        

				
        <p>
          
					I believe preventing cookie usage is not the only way to protect
					your privacy. Instead, you will have full control over all the services
					that are loaded to control any data exchange to foreign services.
				
        </p>
        

				
        <form
          class="details"
        >
          
					
          <dl>
            
						
            <dt>
              
							
              <label>
                
								
                <input
                  name="privacy"
                  type="checkbox"
                  value="analytics"
                />
                
								Google Analytics
							
              </label>
              
						
            </dt>
            
						
            <dd>
              
							Google Analytics is a web analytics service offered by Google that
							tracks and reports website traffic. I track page views and clicks
							on links. The service is running using the "anonymizeIp" setting.
							Read more about 
              <a
                href="https://policies.google.com/privacy?hl=en#footnote-link-info"
                rel="noopener noreferrer"
                target="_blank"
              >
                data privacy
              </a>
              
							from Google Analytics.
						
            </dd>
            

						
            <dt>
              
							
              <label>
                
								
                <input
                  name="privacy"
                  type="checkbox"
                  value="logrocket"
                />
                
								Logrocket
							
              </label>
              
						
            </dt>
            
						
            <dd>
              
							LogRocket is a cloud-based and on-premise logging and session replay
							platform for JavaScript applications, which helps me track UX problems
							and analyze as well as remediate the root causes of bugs.
							Read more about 
              <a
                href="https://logrocket.com/privacy/"
                rel="noopener noreferrer"
                target="_blank"
              >
                privacy policy
              </a>
              
							from LogRocket.
						
            </dd>
            
					
          </dl>
          

					
          <button
            class="allow-selection"
            role="button"
          >
            Use selection
          </button>
          
				
        </form>
        
			
      </details>
      
		
    </div>
    
	
  </div>
  

</div>
`;
